File: src/services/insightsService.ts
import axios from 'axios';
import logger from '../utils/logger';
import { config } from '../config';
import { redis } from '../utils/redis';
import { toTitleCase } from '../utils/transform';

interface FetchInsightsParams {
  dataset: string;
  filters: Record<string, any>;
  limit: number;
  page: number;
}

export const fetchInsights = async ({ dataset, filters, limit, page }: FetchInsightsParams) => {
  // Hardcoded fields per dataset
  let defaultSelect: string[] = [];
  if (dataset) {
    switch (dataset.toLowerCase()) {
      case 'enrolment':
        defaultSelect = ['date', 'state', 'district', 'pincode', 'age_0_5', 'age_5_17', 'age_18_greater'];
        break;
      case 'demographic':
        defaultSelect = ['date', 'state', 'district', 'pincode', 'demo_age_5_17', 'demo_age_17_'];
        break;
      case 'biometric':
        defaultSelect = ['date', 'state', 'district', 'pincode', 'bio_age_5_17', 'bio_age_17_'];
        break;
      default:
        defaultSelect = [];
    }
  }

  // Construct Cache Key
  const stableDynamicKey = JSON.stringify(
    Object.keys(filters)
      .sort()
      .reduce((acc: any, key) => {
        acc[key] = filters[key];
        return acc;
      }, {}),
  );

  const selectFieldsId = defaultSelect.length > 0 ? defaultSelect.sort().join(',') : 'all';
  const cacheKey = `insight:${dataset}:${page}:${limit}:${selectFieldsId}:${stableDynamicKey}`;

  // 1. Check Redis (L2 Cache)
  try {
    const cachedResult: any = await redis.get(cacheKey);
    if (cachedResult) {
      logger.info(`Serving from Redis Cache: ${cacheKey}`);
      return { ...cachedResult, meta: { ...cachedResult.meta, source: 'cache', from_cache: true } };
    }
  } catch (redisErr) {
    logger.warn('Redis Cache Error', redisErr);
  }

  // Map Resource ID
  let resourceId;
  switch (dataset.toLowerCase()) {
    case 'enrolment':
      resourceId = config.resources.enrolment;
      break;
    case 'demographic':
      resourceId = config.resources.demographic;
      break;
    case 'biometric':
      resourceId = config.resources.biometric;
      break;
    default:
      throw new Error('Invalid dataset type.');
  }

  const offset = (Number(page) - 1) * Number(limit);

  // 2. Fetch from External API (L3 Source)
  const apiUrl = `https://api.data.gov.in/resource/${resourceId}`;
  const apiParams: any = {
    'api-key': config.dataGovApiKey,
    format: 'json',
    limit: limit,
    offset: offset,
  };

  // Filter Normalization
  Object.keys(filters).forEach((key: string) => {
    let value = filters[key];
    if (key.toLowerCase() === 'state' || key.toLowerCase() === 'district') {
      if (typeof value === 'string') {
        value = toTitleCase(value);
      }
    }
    apiParams[`filters[${key}]`] = value;
  });

  logger.info(`Fetching from Data.gov.in: ${dataset}`, { apiParams });

  const response = await axios.get(apiUrl, { params: apiParams });
  const data = response.data;

  if (data.status !== 'ok') {
    throw new Error(data.message || 'Error fetching from Source API');
  }

  const fields = data.field ? data.field.map((f: any) => f.id) : [];

  // Apply INTERNAL field selection
  let records = data.records;
  if (defaultSelect.length > 0) {
    records = records.map((record: any) => {
      const filtered: any = {};
      defaultSelect.forEach((f: string) => {
        if (record[f] !== undefined) filtered[f] = record[f];
      });
      return filtered;
    });
  }

  const responsePayload = {
    meta: {
      dataset,
      total: data.total,
      page,
      limit,
      from_cache: false,
      fields: fields,
      source: 'api',
    },
    data: records,
  };

  // 3. Store in Redis
  try {
    await redis.set(cacheKey, responsePayload, { ex: 3600 * 24 }); // Cache for 24 hours
  } catch (err) {
    logger.warn('Failed to set Redis cache', err);
  }

  return responsePayload;
};

File: app/services/insights_service.py
import httpx
import json
from app.core.config import settings
from app.utils.logger import get_logger
from app.utils.redis_client import redis_client

logger = get_logger()

async def fetch_insights(dataset: str, filters: dict, limit: int, page: int):
    # Hardcoded fields
    default_select = []
    ds_lower = dataset.lower()
    
    if ds_lower == 'enrolment':
        default_select = ['date', 'state', 'district', 'pincode', 'age_0_5', 'age_5_17', 'age_18_greater']
    elif ds_lower == 'demographic':
        default_select = ['date', 'state', 'district', 'pincode', 'demo_age_5_17', 'demo_age_17_']
    elif ds_lower == 'biometric':
        default_select = ['date', 'state', 'district', 'pincode', 'bio_age_5_17', 'bio_age_17_']
    
    # Cache Key Construction (Stable)
    # Sort keys for stability
    stable_filters = {k: filters[k] for k in sorted(filters.keys())}
    stable_dynamic_key = json.dumps(stable_filters, separators=(',', ':'))
    
    select_fields_id = ",".join(sorted(default_select)) if default_select else 'all'
    cache_key = f"insight:{dataset}:{page}:{limit}:{select_fields_id}:{stable_dynamic_key}"
    
    # 1. Check Redis (L2)
    try:
        cached = redis_client.get(cache_key)
        if cached:
            if isinstance(cached, str):
                cached = json.loads(cached)
            logger.info(f"Serving from Redis Cache: {cache_key}")
            cached['meta']['source'] = 'cache'
            cached['meta']['from_cache'] = True
            return cached
    except Exception as e:
        logger.warning(f"Redis Cache Error: {str(e)}")
        
    # Map Resource ID
    resource_id = settings.RESOURCES.get(ds_lower)
    if not resource_id:
        raise ValueError("Invalid dataset type.")
        
    offset = (page - 1) * limit
    
    # 2. External API
    api_url = f"https://api.data.gov.in/resource/{resource_id}"
    params = {
        "api-key": settings.DATA_GOV_API_KEY,
        "format": "json",
        "limit": limit,
        "offset": offset
    }
    
    for k, v in filters.items():
        val = v
        # Normalization for State/District to Title Case
        if k.lower() in ['state', 'district'] and isinstance(val, str):
            val = val.title() # Python title() is slightly different but close enough for this
        params[f"filters[{k}]"] = val
        
    logger.info(f"Fetching from Data.gov.in: {dataset} params={params}")
    
    async with httpx.AsyncClient() as client:
        resp = await client.get(api_url, params=params)
        if resp.status_code != 200:
             raise ValueError(f"External API Error: {resp.status_code}")
        
        data = resp.json()
        
    if data.get("status") != "ok":
        raise ValueError(data.get("message") or "Error fetching from Source API")
        
    fields = [f['id'] for f in data.get('field', [])]
    records = data.get('records', [])
    
    # Internal Field Selection
    if default_select:
        new_records = []
        for r in records:
            filtered = {f: r[f] for f in default_select if f in r}
            new_records.append(filtered)
        records = new_records
        
    response_payload = {
        "meta": {
            "dataset": dataset,
            "total": data.get("total"),
            "page": page,
            "limit": limit,
            "from_cache": False,
            "fields": fields,
            "source": "api"
        },
        "data": records
    }
    
    # 3. Store in Redis
    try:
        # Upstash-redis-py: set(name, value, ex)
        # It handles dict serialization? Usually better to serialize.
        # But let's see library docs. It usually accepts string.
        redis_client.set(cache_key, json.dumps(response_payload), ex=86400)
    except Exception as e:
        logger.warning(f"Failed to set Redis cache: {str(e)}")
        
    return response_payload
